crates/matrix-sdk-crypto/src/identities/manager.rs

```diff
@@ -1019,6 +1040,126 @@ impl IdentityManager {
             _ => Ok(None),
         }
     }
+
+    /// Given a list of changed devices, update any [`InboundGroupSession`]s
+    /// which were sent from those devices and which do not have complete
+    /// sender data.
+    ///
+    /// Returns a list of modified InboundGroupSessions which need storing in
+    /// the database.
+    async fn update_sender_data_from_device_changes(
+        &self,
+        device_changes: &DeviceChanges,
+    ) -> Result<Vec<InboundGroupSession>, CryptoStoreError> {
+        let mut result = Vec::new();
+
+        for device in device_changes.new.iter().chain(device_changes.changed.iter()) {
+            // 1. Look for InboundGroupSessions from the device whose sender_data is
+            //    UnknownDevice. For such sessions, we now have the device, and can update
+            //    the sender_data accordingly.
+            //
+            // In theory, we only need to do this for new devices. In practice, I'm a bit
+            // worried about races leading us to getting stuck in the
+            // UnknownDevice state, so we'll paper over that by doing this check
+            // on device updates too.
+            let mut result1 = self
+                .update_sender_data_for_sessions_for_device(device, SenderDataType::UnknownDevice)
+                .await?;
+            result.append(&mut result1);
+
+            // 2. If, and only if, the device is now correctly cross-signed (ie,
+            //    device.is_cross_signed_by_owner() is true, and we have the master
+            //    cross-signing key for the owner), look for InboundGroupSessions from the
+            //    device whose sender_data is DeviceInfo. We can also update the sender_data
+            //    for these sessions.
+            //
+            // In theory, we can skip a couple of steps of the SenderDataFinder algorithm,
+            // because we're doing the cross-signing check here. In practice,
+            // it's *way* easier just to use the same logic.
+            let device_owner_identity = self.store.get_user_identity(device.user_id()).await?;
+            if device_owner_identity.is_some_and(|id| device.is_cross_signed_by_owner(&id)) {
+                let mut result1 = self
```

Comment:
`result1` -> `sessions_device_info` or similar

Reply:
removed as a side-effect of 74e63219d